kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -163,6 +192,21 @@ public Transaction build(Engine engine) {
       }
     }
 
+    /* --------------- Domain Metadata Protocol upgrade if necessary------------ */
+    if (!TableFeatures.isDomainMetadataSupported(protocol)) {
+      if (!domainMetadatasAdded.isEmpty()) {
+        // This txn is setting a domain metadata, enable the feature in the protocol
+        logger.info(
+            "Automatically enabling writer feature: {}", DOMAIN_METADATA_W_FEATURE.featureName());
+        protocol =
+            protocol.withNewWriterFeatures(
+                Collections.singleton(DOMAIN_METADATA_W_FEATURE.featureName()));
+        shouldUpdateProtocol = true;
+      }
+      // If domainMetadatasRemoved is non-empty we do nothing. A DomainDoesNotExistException will be
+      // thrown in `getDomainMetadatasToCommit` since the domain cannot exist in the readSnapshot.
+    }
+
```

Comment:
Synced with @vkorukanti; we will merge this PR as is and in https://github.com/delta-io/delta/pull/4159 we will integrate this into one single protocol upgrade by having `autoUpgradeProtocolBasedOnMetadata` take an argument like `addsDomainMetadatas`